File: ares/reference/swift-ui/mac_sources/ARES-Mac/FaceState.swift
import SwiftUI

enum FaceState: Equatable, CustomStringConvertible {
    case idle
    case awakened
    case listening
    case thinking
    case speaking
    case sleeping
    
    var color: Color {
        switch self {
        case .idle: return Color(red: 0.2, green: 0.6, blue: 1.0)
        case .awakened: return Color(red: 0.3, green: 0.8, blue: 1.0)
        case .listening: return Color(red: 0.4, green: 0.9, blue: 0.5)
        case .thinking: return Color(red: 1.0, green: 0.7, blue: 0.2)
        case .speaking: return Color(red: 0.8, green: 0.4, blue: 1.0)
        case .sleeping: return Color(red: 0.1, green: 0.2, blue: 0.4)
        }
    }
    
    var opacity: Double {
        switch self {
        case .idle: return 0.3
        case .awakened: return 0.6
        case .listening: return 0.7
        case .thinking: return 0.8
        case .speaking: return 0.9
        case .sleeping: return 0.15
        }
    }
    
    var pulseSpeed: Double {
        switch self {
        case .idle: return 1.5
        case .sleeping: return 0.5
        case .thinking: return 3.0
        default: return 2.0
        }
    }
    
    var pulseAmount: Double {
        switch self {
        case .idle: return 3
        case .sleeping: return 2
        case .speaking: return 5
        default: return 4
        }
    }
    
    var pupilOffset: Double {
        switch self {
        case .listening: return 2
        case .thinking: return -2
        case .speaking: return 0
        default: return 0
        }
    }
    
    var description: String {
        switch self {
        case .idle: return "Idle"
        case .awakened: return "Awake"
        case .listening: return "Listening"
        case .thinking: return "Thinking"
        case .speaking: return "Speaking"
        case .sleeping: return "Sleeping"
        }
    }
}

File: ares/core/face_state.py
"""ARES face state machine — maps cognition to expression.

6 states, each with visual parameters for color, opacity, pulse speed, pupil offset.
Directly usable by both Python cognition layer and SwiftUI FaceRenderer.
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum
from typing import Tuple


class FaceState(Enum):
    """Emotional/operational states for the companion face."""

    IDLE = "idle"  # Waiting, ambient awareness
    AWAKENED = "awakened"  # Wake word detected, attention focused
    LISTENING = "listening"  # Processing speech input
    THINKING = "thinking"  # LLM inference in progress
    SPEAKING = "speaking"  # TTS output, mouth animation
    SLEEPING = "sleeping"  # Dormant / low-power


@dataclass(frozen=True)
class FaceConfig:
    """Visual parameters for a face state."""

    color: Tuple[float, float, float]  # RGB 0-1
    opacity: float  # 0-1
    pulse_speed: float  # animation cycles/sec
    pulse_amount: float  # 0-1 intensity
    pupil_offset: Tuple[float, float]  # x,y offset from center (-1 to 1)


STATE_CONFIGS: dict[FaceState, FaceConfig] = {
    FaceState.IDLE: FaceConfig(
        color=(0.6, 0.6, 0.7), opacity=0.6, pulse_speed=0.3, pulse_amount=0.1, pupil_offset=(0.0, 0.0)
    ),
    FaceState.AWAKENED: FaceConfig(
        color=(0.4, 0.7, 1.0), opacity=0.9, pulse_speed=1.5, pulse_amount=0.2, pupil_offset=(0.0, -0.1)
    ),
    FaceState.LISTENING: FaceConfig(
        color=(0.3, 0.9, 0.5), opacity=1.0, pulse_speed=2.0, pulse_amount=0.3, pupil_offset=(0.05, 0.05)
    ),
    FaceState.THINKING: FaceConfig(
        color=(0.9, 0.6, 0.2), opacity=1.0, pulse_speed=3.0, pulse_amount=0.4, pupil_offset=(0.0, 0.2)
    ),
    FaceState.SPEAKING: FaceConfig(
        color=(0.5, 0.8, 1.0), opacity=1.0, pulse_speed=0.5, pulse_amount=0.05, pupil_offset=(0.0, -0.05)
    ),
    FaceState.SLEEPING: FaceConfig(
        color=(0.2, 0.2, 0.3), opacity=0.3, pulse_speed=0.1, pulse_amount=0.05, pupil_offset=(0.0, 0.1)
    ),
}


def get_face_config(state: FaceState) -> FaceConfig:
    """Get visual config for a face state."""
    return STATE_CONFIGS[state]


def emotion_to_face_state(emotion: str, is_processing: bool = False) -> FaceState:
    """Map a high-level emotion string to a face state.

    Called by the cognition layer when ARES determines what to express.
    """
    if emotion in ("happy", "excited", "enthusiastic"):
        return FaceState.AWAKENED
    elif emotion in ("sad", "disappointed", "concerned"):
        return FaceState.IDLE
    elif emotion in ("thinking", "analyzing", "computing"):
        return FaceState.THINKING
    elif emotion in ("surprised", "curious", "interested"):
        return FaceState.LISTENING
    elif emotion in ("neutral", "calm", "focused"):
        return FaceState.IDLE

    return FaceState.THINKING if is_processing else FaceState.IDLE
